### ISSUE TITLE
Unable to set paged results response

### ISSUE THREAD

**Issue by andrius-kulbis** [ldaptools/ldaptools#17]
I have the followinf setup:

```
 $domain = (new DomainConfiguration('ldap'))
     ->setServers(SERVERS)
     ->setBaseDn(BASE_DN)
     ->setUsername(USER)
     ->setPassword(PASSWORd)
     ->setUsePaging(true)
     ->setPageSize(5)
     ->setLdapType(LdapConnection::TYPE_OPENLDAP)
;

$ldapManager = new LdapManager(new Configuration($domain));

$users = $ldapManager->buildLdapQuery()
    ->getLdapQuery()
    ->getResult()
;
```
This setup gives me  
```
LdapTools\Exception\LdapConnectionException: Unable to set paged results response: Success in /project/vendor/ldaptools/ldaptools/src/LdapTools/Connection/PageControl.php:126
```

I thought that the problem is ldap not supporting the paged results. But the following code works as expected:
```
$connection = ldap_connect(HOST);

ldap_set_option($connection, LDAP_OPT_PROTOCOL_VERSION, 3);
ldap_bind($connection, USER, PASS);

$cookie = '';
do {
    ldap_control_paged_result($connection, 5, false, $cookie);

    $result = ldap_search($connection, $dn, "(objectClass=eduPerson)", $attributes);
    $entries = ldap_get_entries($connection, $result);

    foreach ($entries as $entry) {
        echo $entry['dn'] . "<br>";
    }

    ldap_control_paged_result_response($connection, $result, $cookie);
} while ($cookie !== null && $cookie !== '');
```

Am I missing something, or there is a problem with paged results in `ldaptools`?

**Comment by ChadSikorra** [comment]
Weird. If you comment out the code that throws the exception on line 126 of `LdapTools\Connection\PageControl` does the query work? It seems weird that it returns false for updating paged results response. I haven't tested in OpenLDAP in a while though. I can verify paging works fine for AD. I can test more later.

On a side note, I think there's actually a bug with the way LDAP paging is implemented in PHP. As I have noticed issues when attempting to reset the page control after the query is complete (then certain queries don't work afterwards).

**Comment by andrius-kulbis** [comment]
Yes, I found the problem. Somehow the OpenLDAP fails the query if there is no filter assigned. So instead of:
```
$users = $ldapManager->buildLdapQuery()
    ->getLdapQuery()
    ->getResult()
;
```

I need to supply atleast a basic filter:
```
$users = $ldapManager->buildLdapQuery()
    ->where(['objectClass' => 'eduPerson'])
    ->getLdapQuery()
    ->getResult()
;
```

**Comment by ChadSikorra** [comment]
Ah, yes. That would do it. I overlooked that in your example. If you don't specify a schema type using the from method then you need to specify at least some basic filter. This would be true in any LDAP implementation (OpenLDAP or AD).

What I should do is give a more helpful error message before it gets that far that an empty filter is not valid.

**Comment by ChadSikorra** [comment]
The query will return all of the results at once. It's not implemented to do interactive paging, but rather to adhere to paging restrictions against the LDAP server so it doesn't refuse the query when too many results would be returned.

However, I'm not opposed to implementing some sort of interactive paging mechanism like what you're looking for. But I'd have to think about how to implement it.

**Comment by andrius-kulbis** [comment]
That would be awesome. Listing users is a required feature in an application I'm building and it is not practical loading all of them at once.

**Comment by ChadSikorra** [comment]
You can set a limit to how many are loaded at once by using `setSizeLimit($size);`:

```php
$users = $ldapManager->buildLdapQuery()
    ->where(['objectClass' => 'eduPerson'])
    ->setSizeLimit(100)
    ->getLdapQuery()
    ->getResult()
;
```

**Comment by ChadSikorra** [comment]
The only other option I can think of is to add a `startAt($position)` method to the query builder. Then the paged results could be started at a certain page set. That along with the max size and page size would give enough control to get chunks of data at a time on each query.

The real problem is that LDAP paging does not provide any control over an offset when retrieving data. So it always starts with the first result set and works it's way up on each response. So with the above solution I'd basically just be discarding certain result sets until we get to the offset. 

Another solution would be the [LDAP_CONTROL_VLVREQUEST](https://msdn.microsoft.com/en-us/library/aa366178%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396). Which seems to be a much enhanced paging-type of operation. But I'm pretty sure I'd run into all sorts of issues getting that working with PHPs LDAP extension. It would probably require built in functions into the PHP LDAP extension just like the paging functions. The sad part about this is this control somehow never made it as an accepted RFC, and the draft [appears to be sitting in limbo](https://www.ietf.org/archive/id/draft-ietf-ldapext-ldapv3-vlv-09.txt). Pretty weird, considering all major LDAP implementations support it.

**Comment by ChadSikorra** [comment]
Query operations now require a non-empty filter value. If it is empty it will throw a more helpful message. Added with: https://github.com/ldaptools/ldaptools/commit/74e7edecf43058032635e6917641cadb115bc47f. Unfortunately I was abusing that in my specs so I had to go through and fix a bunch of them.

It will take a little while to determine what to do about the paging situation. But if your goal is to limit the results returned I would set a size limit on the query via the method I suggested.

**Comment by andrius-kulbis** [comment]
Well I could use some kind of decorator for the `LdapQuery` in combination with `sizeLimit` to get the pagination feeling without fetching all the records every time.

**Comment by ChadSikorra** [comment]
It's not so easy at the moment, because all of the raw LDAP results are collected in the operation handler before being returned back, then all of the results are normalized/hydrated at once.

**Comment by cooperaj** [comment]
Would like to see this implemented if possible. I'd assumed I would have had some control over the paging after turning it on but this being the case I'll just lock it to a reasonable number.
